### ISSUE TITLE
[BUG] examples/balance/in.balance.bond.*: abort due to bond atoms missing

### ISSUE THREAD

**Issue by evan-charmworks** [lammps/lammps#2630]
**Summary**

Running examples/balance/in.balance.bond.fast and in.balance.bond.slow can lead to an MPI_Abort because bond atoms are missing.

```
$ mpirun -np 4 ../../src/lmp_mpi -in in.balance.bond.fast
LAMMPS (10 Feb 2021)
Lattice spacing in x,y,z = 1.1655347 2.0187654 1.1655347
Created orthogonal box = (0.0000000 0.0000000 -0.58276737) to (58.276737 40.375308 0.58276737)
  2 by 2 by 1 MPI processor grid
Created 361 atoms
  create_atoms CPU = 0.001 seconds
Finding 1-2 1-3 1-4 neighbors ...
  special bond factors lj:    0        1        1       
  special bond factors coul:  0        1        1       
     0 = max # of 1-2 neighbors
     1 = max # of special neighbors
  special bonds CPU = 0.000 seconds
Neighbor list info ...
  update every 1 steps, delay 10 steps, check yes
  max neighbors/atom: 2000, page size: 100000
  master list distance cutoff = 2.8
  ghost atom cutoff = 2.8
  binsize = 1.4, bins = 42 29 1
  2 neighbor lists, perpetual/occasional/extra = 1 1 0
  (1) command create_bonds, occasional
      attributes: full, newton on
      pair build: full/bin
      stencil: full/bin/2d
      bin: standard
  (2) pair lj/cut, perpetual
      attributes: half, newton on
      pair build: half/bin/newton
      stencil: half/bin/2d/newton
      bin: standard
Added 1014 bonds, new total = 1014
Finding 1-2 1-3 1-4 neighbors ...
  special bond factors lj:    0        1        1       
  special bond factors coul:  0        1        1       
     6 = max # of 1-2 neighbors
     6 = max # of special neighbors
  special bonds CPU = 0.000 seconds
Neighbor list info ...
  update every 1 steps, delay 0 steps, check yes
  max neighbors/atom: 2000, page size: 100000
  master list distance cutoff = 2.8
  ghost atom cutoff = 10
  binsize = 1.4, bins = 42 29 1
  1 neighbor lists, perpetual/occasional/extra = 1 0 0
  (1) pair lj/cut, perpetual
      attributes: half, newton on
      pair build: half/bin/newton
      stencil: half/bin/2d/newton
      bin: standard
Setting up Verlet run ...
  Unit style    : lj
  Current step  : 0
  Time step     : 0.005
Per MPI rank memory allocation (min/avg/max) = 5.123 | 5.142 | 5.176 Mbytes
Step Temp E_pair Press f_10[3] f_10 
       0    25.701528   -2.2032569    3.1039469    3.2354571    1.0526316 
     100    27.623422    -6.228166    2.6542136    1.2631579    1.0083102 
     200     33.35302   -15.746749    3.2018248    1.2963989    1.0193906 
     300     39.17734     -24.1557    4.9116986    1.2963989    1.0193906 
     400    41.660701   -27.615203    8.6214679    1.3518006    1.0083102 
     500    37.154928   -24.096946    3.2656176    1.3296399    1.0193906 
     600    35.059869   -21.524213    2.3729549    1.3296399    1.0083102 
     700    36.737496   -22.010963     3.261542    1.3296399    1.0083102 
     800    39.234921   -25.118003    4.7186521    1.4182825    1.0193906 
     900    38.542226   -24.537471    4.8070434    1.3628809    1.0083102 
    1000    38.062705    -25.72569    5.0043027    1.2853186    1.0083102 
    1100    36.140107   -24.848575    5.9104398    1.2077562    1.0193906 
    1200    31.390064   -22.774218    3.6364797    1.2299169    1.0083102 
    1300    38.100741   -21.521967     5.528522    1.1080332    1.0083102 
ERROR on proc 2: Bond atoms 97 135 missing on proc 2 at step 1307 (../ntopo_bond_all.cpp:60)
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 2 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
```

The issue is nondeterministic. It does not always happen, and when it does, the simulation will be in a different place each time.

```
ERROR on proc 1: Bond atoms 308 338 missing on proc 1 at step 1264 (../ntopo_bond_all.cpp:60)
ERROR on proc 3: Bond atoms 310 311 missing on proc 3 at step 1264 (../ntopo_bond_all.cpp:60)
```
```
ERROR on proc 3: Bond atoms 320 321 missing on proc 3 at step 1191 (../ntopo_bond_all.cpp:60)
```
```
ERROR on proc 3: Bond atoms 212 252 missing on proc 3 at step 1227 (../ntopo_bond_all.cpp:60)
```
```
ERROR on proc 3: Bond atoms 211 212 missing on proc 3 at step 1184 (../ntopo_bond_all.cpp:60)
```

**LAMMPS Version and Platform**

Reproduced with both stable_29Oct2020 tag (`LAMMPS (29 Oct 2020)`) and latest git master, f245467f32dba758164067116423e119eb5ec78a (`LAMMPS (10 Feb 2021)`).

My platform:

```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.2 LTS
Release:	20.04
Codename:	focal
$ mpicxx --version
g++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ mpirun --version
mpirun (Open MPI) 4.0.3

Report bugs to http://www.open-mpi.org/community/help/
```

**Expected Behavior**

The example program should complete successfully.

**Actual Behavior**

The manual explains the error message:

> *****Bond atoms %d %d missing on proc %d at step %ld*****
> The second atom needed to compute a particular bond is missing on this processor. Typically this is because the pairwise cutoff is set too short or the bond has blown apart and an atom is too far away.

**Steps to Reproduce**

`cd src && make yes-molecule && make mpi -j4 && cd ../examples/balance && for i in {1..100} ; do echo $i ; mpirun -np 4 ../../src/lmp_mpi -in in.balance.bond.fast || break ; done`

**Further Information, Files, and Links**

I found this thread from the LAMMPS mailing list that suggests this problem is an issue with the examples' problem statements and not a code bug in LAMMPS itself.

https://lammps.sandia.gov/threads/msg01442.html

**Comment by akohlmey** [comment]
This is likely a known problem. If it is, it has nothing to do with the statements from the mailing list.
In this case, the system is too small (or the CPU too fast or both) and the measured times on some machines and MPI libraries don't have a sufficient resolution. This will lead to bogus load balancing constraints which can shrink a subdomain to become too small to include all bonded atoms as ghosts.

**Comment by stanmoore1** [comment]
>This will lead to bogus load balancing constraints which can shrink a subdomain to become too small to include all bonded atoms as ghosts.

Is there no way to add a check for this, or set the minimum box size to a sane value?

**Comment by akohlmey** [comment]
> > This will lead to bogus load balancing constraints which can shrink a subdomain to become too small to include all bonded atoms as ghosts.
> 
> Is there no way to add a check for this, or set the minimum box size to a sane value?

Just checked it out. This is a different scenario from what I suspected. I have an idea for addressing the timing issue but never got around to trying it out. Now may be as good a time as any since I am already chasing down issues with pending PRs and a few items on my TODO list. This example, however, using rcb based on particle counts. That is something for @sjplimp to look into.

**Comment by akohlmey** [comment]
This seems to be resolved. I cannot reproduce it anymore. Was probably fixed by commit de6bae633339edd07a51adbe8de9e2179639bc16
